source/class/zx/utils/AsyncMemoizer.js

```diff
@@ -0,0 +1,180 @@
+   *
+   * @param {Generator?} generator
+   */
+  construct(generator) {
+    super();
+    if (generator) {
+      this.setGenerator(generator);
+    }
+    this.__cleanupTimer = new zx.utils.Timeout(0, () => this.__cleanupExpiredValues());
+    this.bind("cleanupInterval", this.__cleanupTimer, "duration");
+  },
+  properties: {
+    /**
+     * The generator function that will be called to compute the value
+     * @type {Generator}
+     */
+    generator: {
+      check: "Function",
+      nullable: false,
+      event: "changeGenerator"
+    },
+    /**
+     * If non-zero, the cache will expire after this many milliseconds.
+     */
+    expiryTime: {
+      check: "Number",
+      init: 0, // Default to no expiry
+      nullable: false,
+      event: "changeExpiryTime"
+    },
+    /**
+     * How often to check for expired values in the cache.
+     * Only used if expiryTime is non-zero.
+     */
+    cleanupInterval: {
+      check: "Number",
+      init: 10000, // Default to 10 seconds
+      nullable: false,
+      event: "changeCleanupInterval"
+    }
+  },
+  members: {
+    /**
+     * @type {Map<InputData, OutputData>}
+     * We have to use a Map for this and not a POJO because otherwise we would only be limited to string-y keys.
+     */
+    __cache: new Map(),
+
+    /**
+     * @type {zx.utils.Timeout}
+     */
+    __cleanupTimer: null,
+
+    /**
+     * Get the value for the given input, computing it if necessary.
+     * If the value is being computed, it will return a promise that resolves to the value.
+     *
+     * @param {InputData} input
+     * @return {Promise<OutputData> | OutputData}
+     */
+    get(input) {
+      let cache = this.__cache;
+      let generator = this.getGenerator();
+
+      if (!cache.has(input)) {
+        let record = {};
+
+        /**
+         *
+         * @param {OutputData} value
+         * @returns {OutputData}
+         */
+        const onReady = value => {
+          record.value = value;
+          delete record.promise;
+          record.lastActive = Date.now();
+          this.__cleanupTimer.setEnabled(true); // Ensure the cleanup timer is running
+          return value;
+        };
+
+        let out = generator(input);
+        if (qx.Promise.isPromise(out)) {
+          record.promise = out.then(onReady).catch(e => {
+            this.__cache.delete(input); // Remove from cache if promise rejected
+            throw e;
+          });
+        } else {
+          onReady(out);
+        }
+
+        this.__cache.set(input, record);
+        return out;
+      } else {
+        let record = this.__cache.get(input);
+        if (record.value !== undefined) {
+          return record.value;
+        }
+        return record.promise;
+      }
+    },
+
+    /**
+     * Kicks the cache entry for the given input, updating its last active time.
+     * This is useful to keep the entry alive if it is still in use.
+     * @param {InputData} input
+     */
+    kick(input) {
```

Comment:
This is not a good name either - "kick" could mean any number of things, eg restart or remove.  A better name would be "updateLastActive", or "keepAlive"